### PR TITLE
[sgd] Document and add simple MNIST example

### PR DESCRIPTION
**Pull request by ericl**
## Related issue number

https://github.com/ray-project/ray/issues/3138

### CODE REVIEW

**Comment by ericl** [comment]
The tune test here passes after https://github.com/ray-project/ray/pull/3244

File: doc/source/distributed_sgd.rst [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+    sgd = DistributedSGD(
+        model_creator,
+        num_workers=2,
+        devices_per_worker=4,
+        gpu=True,
+        strategy="ps")
+
+    for i in range(NUM_ITERS):
+        print(sgd.step(fetch_stats=True))
```
Comment (richardliaw) [inline doc/source/distributed_sgd.rst]: maybe not `fetch_stats` all the time in the doc example?

File: python/ray/experimental/sgd/sgd_worker.py [hunk 2]
```diff
@@ -182,15 +182,28 @@ def __init__(self,
...
+        return feed_dict
+
     def foreach_model(self, fn):
-        return [fn(m) for m in self.models]
+        with self.sess.as_default():
+            return [fn(m) for m in self.models]
 
     def foreach_worker(self, fn):
-        return fn(self)
+        with self.sess.as_default():
```
Comment (richardliaw) [inline python/ray/experimental/sgd/sgd_worker.py]: what was the issue before (for all the changes in this file?)

**Review by richardliaw — COMMENTED** [review]
Looks good, just want to know what's going on

File: doc/source/distributed_sgd.rst (hunk 1 above)
Comment (richardliaw) [inline doc/source/distributed_sgd.rst]:
```suggestion
        sgd.step()
```

**Review by richardliaw — APPROVED** [review]

**Review by robertnishihara — APPROVED** [review]

**Review by pcmoritz — APPROVED** [review]

**Comment by pcmoritz** [comment]
Small linting problems:
```
./python/ray/experimental/sgd/mnist_example.py:50:13: F841 local variable 'e' is assigned to but never used
./python/ray/experimental/sgd/mnist_example.py:53:55: F821 undefined name 'e'
```

**Comment by pcmoritz** [comment]
Unfortunately the lint god doesn't seem to be appeased yet :(
```
./python/ray/experimental/sgd/mnist_example.py:51:13: F841 local variable 'e' is assigned to but never used
```

**Comment by pcmoritz** [comment]
Hmm, it looks like in Jenkins, some worker tries to use the plasma op before it is finished compiling, am I interpreting this correctly?

**Comment by richardliaw** [comment]
And travis is just stalling?

**Comment by pcmoritz** [comment]
jenkins retest this please

**Comment by ericl** [comment]
jenkins retest this please

**Comment by ericl** [comment]
Seems the tests OOM when attempting to evaluate test accuracy (perhaps memory related). Trying with those removed.

**Pull request merged**